src/bentoml/_internal/io_descriptors/utils.py

```diff
@@ -1,68 +1,97 @@
+            line = buffer.readline()
+            if not line:
+                break
+            if line == "\n":
+                break
+            if line.startswith("data: "):
+                if first_data_line:
+                    first_data_line = False
+                    event["data"] = line[6:-1]
+                else:
+                    if data_buffer is None:
+                        # only init data_buffer when there is more than one data line
+                        data_buffer = io.StringIO()
+                        data_buffer.write(event.get("data", ""))
+                    data_buffer.write("\n")
+                    data_buffer.write(line[6:-1])
+            if line.startswith("event: "):
+                event["event"] = line[7:].strip()
+            if line.startswith("id: "):
+                event["id"] = line[4:].strip()
+            if line.startswith("retry: "):
+                event["retry"] = int(line[7:].strip())
 
-    @staticmethod
-    async def from_iterator(
-        async_iterator: AsyncIterator[str],
-    ) -> AsyncGenerator["SSE", None]:
-        buffer = io.StringIO()
-        async for chunk in async_iterator:
-            buffer.write(chunk)
-            # Move to the beginning to read content
-            buffer.seek(0, io.SEEK_SET)
-            content = buffer.read()
+        if data_buffer is not None:
+            event["data"] = data_buffer.getvalue()
+            data_buffer.close()
 
-            while "\n\n" in content:
-                event_text, _, remaining = content.partition("\n\n")
-                fields = {"data": "", "event": None, "id": None, "retry": None}
-                data = []
-                for line in event_text.split("\n"):
-                    key, _, value = line.partition(": ")
-                    if key == "data":
-                        data.append(value)
-                    elif key == "retry":
-                        fields[key] = int(value)
-                    elif key in fields:
-                        fields[key] = value
-                    else:
-                        raise BentoMLException("Invalid SSE message")
-                fields["data"] = "\n".join(data)
-                yield SSE(**fields)
+        return SSE(**event)
 
-                # Reset buffer
-                buffer.seek(0)
-                buffer.truncate()
-                buffer.write(remaining)
-                content = remaining
+    @classmethod
+    async def from_iterator(
+        cls,
+        async_iterator: t.AsyncIterator[str],
+    ) -> t.AsyncGenerator[SSE, None]:
+        with io.StringIO() as buffer:
+            read_cursor = 0
+            last_chunk = ""
+            async for chunk in async_iterator:
+                buffer.write(chunk)
+                if (
+                    "\n\n" in chunk
+                    or chunk
+                    and chunk[0] == "\n"
+                    and last_chunk
+                    and last_chunk[-1] == "\n"
```

Comment:
Is this part needed? Could we always just append the buffer until we see `\n\n` and then update the cursor?

Reply:
Chunking may happens right in the meddle of `\n\n`

Reply:
Yeah, if that happens maybe we just simply wait for the next chunk to arrive?

Reply:
what do you mean?

Reply:
I don't think so. You can take a closer look at the process here.

Reply:
Right, i understood what you wrote. I think we can just have `while '\n\n' in chunks` condition, and that will solve two problem at once:

1. The scenario you mentioned, where its chunked between two `\n` character, it will simply wait for the next chunk to enter the buffer then we will have the full delimiter again.
2. Multiple events in the same chunk, it will then yield multiple times before it receives the next chunk into buffer